pkg/apis/certmanager/v1beta1/const.go

```diff
@@ -0,0 +1,43 @@
+/*
+Copyright 2019 The Jetstack cert-manager contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import "time"
+
+const (
+	// minimum permitted certificate duration by cert-manager
+	MinimumCertificateDuration = time.Hour
```

Comment:
This is funny.... is this enforced somewhere? I remember being able to set it lower than this (which caused a disaster), i don't remember it being enforced in the validating webhook.

Reply:
That's a good question.. I'm actually not too sure. I think it may not be, and it's also quite a difficult thing for us to *enforce* given the CA can override the behaviour.

All of cert-manager code needs to tolerate _any_ duration, and can't really rely on this being true.

That said, supposing a CA returns a Certificate valid for 1s, or a user requests a certificate valid for 1s, what would we really want to happen here? That's an excessively short duration, and will lead to cert-manager performing _lots_ of certificate requests.

As a cluster administrator, I'd like to have _some_ guard against this. Perhaps a flag on the controller which is a universal minimum value? (could be enforced by the webhook then?).

How we do handle the case where a user *doesn't* explicitly set a duration, yet the CA still returns one valid for 1s? Should we go into that tight request loop *despite* the minimum duration not being met?

Thinking even _further_.. what if a CA returns a certificate that is already expired?

Reply:
It is used in the validation functions:

```
git grep MinimumCertificateDuration
pkg/apis/certmanager/v1alpha2/const.go: MinimumCertificateDuration = time.Hour
pkg/apis/certmanager/v1alpha3/const.go: MinimumCertificateDuration = time.Hour
pkg/internal/apis/certmanager/validation/certificate.go:        if duration < cmapiv1alpha2.MinimumCertificateDuration {
pkg/internal/apis/certmanager/validation/certificate.go:                el = append(el, field.Invalid(fldPath.Child("duration"), duration, fmt.Sprintf("certificate duration must be greater than %s", cmapiv1alpha2.MinimumCertificateDuration)))
pkg/internal/apis/certmanager/validation/certificate_test.go:                   errs: []*field.Error{field.Invalid(fldPath.Child("duration"), usefulDurations["half hour"].Duration, fmt.Sprintf("certificate duration must be greater than %s", cmapiv1alpha2.MinimumCertificateDuration))},

```

...but I also remember setting duration lower than this in the nginx workshop, so is the validation broken?

Or, perhaps we never set the duration and relied instead on the TPP policy to set a duration.
 * https://github.com/jetstack/cert-manager-nginx-plus-lab/blob/master/pingpong.yaml

Reply:
Opened https://github.com/jetstack/cert-manager/issues/3067 to track this

Reply:
@wallrj venafi doesn't support setting the duration so yes we did there, however i tested this with the CA issuer before. Was before experimental controllers so cannot tell how it behaves now.